File: src/main/java/io/cloudflight/architecture/structurizr/internal/service/export/StructurizrWorkspaceExportService.java
package io.cloudflight.architecture.structurizr.internal.service.export;

import com.structurizr.Workspace;
import com.structurizr.api.StructurizrClientException;
import com.structurizr.api.WorkspaceApiClient;
import io.cloudflight.architecture.structurizr.WorkspaceExportService;
import io.cloudflight.architecture.structurizr.autoconfigure.StructurizrProperties;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nonnull;

public final class StructurizrWorkspaceExportService implements WorkspaceExportService {

    private final StructurizrProperties properties;
    private final WorkspaceApiClient structurizrClient;

    private static final Logger LOG = LoggerFactory.getLogger(StructurizrWorkspaceExportService.class);

    public StructurizrWorkspaceExportService(StructurizrProperties properties, WorkspaceApiClient structurizrClient) {
        this.properties = properties;
        this.structurizrClient = structurizrClient;
    }

    @Override
    public void exportWorkspace(@Nonnull Workspace workspace) {
        if (properties.export().structurizr().enabled()) {
            try {
                LOG.info("Exporting workspace to ID {}", properties.export().structurizr().id());
                structurizrClient.putWorkspace(properties.export().structurizr().id(), workspace);
            } catch (StructurizrClientException e) {
                LOG.error("Could not put workspace.", e);
                throw new IllegalArgumentException("Could not put workspace", e);
            }
        }
    }
}

File: src/main/kotlin/io/cloudflight/architecture/structurizr/service/export/StructurizrWorkspaceExportService.kt
package io.cloudflight.architecture.structurizr.service.export

import com.structurizr.Workspace
import com.structurizr.api.StructurizrClient
import com.structurizr.api.StructurizrClientException
import io.cloudflight.architecture.structurizr.WorkspaceExportService
import io.cloudflight.architecture.structurizr.autoconfigure.StructurizrProperties
import org.slf4j.LoggerFactory

internal class StructurizrWorkspaceExportService(
    private val properties: StructurizrProperties,
    private val structurizrClient: StructurizrClient
) : WorkspaceExportService {
    override fun exportWorkspace(workspace: Workspace) {
        if (properties.export.structurizr.enabled) {
            try {
                LOG.info("Exporting workspace to ID " + properties.export.structurizr.id + " at " + structurizrClient.url)
                structurizrClient.putWorkspace(properties.export.structurizr.id, workspace)
            } catch (e: StructurizrClientException) {
                LOG.error("Could not put workspace.", e)
                throw IllegalArgumentException("Could not put workspace", e)
            }
        }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(StructurizrWorkspaceExportService::class.java)
    }
}